### PR TITLE
Update the service to std futures

### PR DESCRIPTION
**Pull request by expenses**
Part of #3099.

### CODE REVIEW

File: client/cli/src/informant.rs [hunk 1]
```diff
@@ -17,7 +17,7 @@
 //! Console informant. Prints sync progress and block events. Runs on the calling thread.
 
 use sc_client_api::BlockchainEvents;
-use futures::{StreamExt, TryStreamExt, FutureExt, future, compat::Stream01CompatExt};
+use futures::{StreamExt, FutureExt, future};
```
Comment (tomaka) [inline client/cli/src/informant.rs]:
```suggestion
use futures::prelude::*;
```

File: client/consensus/aura/src/lib.rs [hunk 2]
```diff
@@ -148,7 +148,7 @@ pub fn start_aura<B, C, SC, E, I, P, SO, CAW, Error, H>(
 	force_authoring: bool,
 	keystore: KeyStorePtr,
 	can_author_with: CAW,
-) -> Result<impl futures01::Future<Item = (), Error = ()>, sp_consensus::Error> where
+) -> Result<impl futures::Future<Output = ()>, sp_consensus::Error> where
```
Comment (tomaka) [inline client/consensus/aura/src/lib.rs]: The style was `futures01::` in order to make it obvious that we have a futures 0.1. Now that there's only a single future type, we should do some nit-picky styling.

```suggestion
) -> Result<impl Future<Output = ()>, sp_consensus::Error> where
```

File: client/service/src/chain_ops.rs [hunk 3]
```diff
@@ -69,7 +66,7 @@ impl<
 		self,
 		input: impl Read + Seek + Send + 'static,
 		force: bool,
-	) -> Box<dyn Future<Item = (), Error = Error> + Send> {
+	) -> Box<dyn Future<Output = Result<(), Error>> + Send + Unpin> {
```
Comment (tomaka) [inline client/service/src/chain_ops.rs]:
```suggestion
	) -> Pin<Box<dyn Future<Output = Result<(), Error>> + Send>> {
```

File: client/service/src/chain_ops.rs [hunk 4]
```diff
@@ -206,7 +203,7 @@ impl<
 				return std::task::Poll::Pending;
 			}
 		});
-		Box::new(import.compat())
+		Box::new(import)
```
Comment (tomaka) [inline client/service/src/chain_ops.rs]:
```suggestion
		Box::pin(import)
```

File: client/service/src/chain_ops.rs [hunk 5]
```diff
@@ -215,7 +212,7 @@ impl<
 		from: NumberFor<TBl>,
 		to: Option<NumberFor<TBl>>,
 		json: bool
-	) -> Box<dyn Future<Item = (), Error = Error>> {
+	) -> Box<dyn Future<Output = Result<(), Error>> + Unpin> {
```
Comment (tomaka) [inline client/service/src/chain_ops.rs]: Same

File: client/service/src/chain_ops.rs [hunk 6]
```diff
@@ -296,7 +293,7 @@ impl<
 	fn check_block(
 		self,
 		block_id: BlockId<TBl>
-	) -> Box<dyn Future<Item = (), Error = Error> + Send> {
+	) -> Box<dyn Future<Output = Result<(), Error>> + Send + Unpin> {
```
Comment (tomaka) [inline client/service/src/chain_ops.rs]: Same

```suggestion
	) -> Pin<Box<dyn Future<Output = Result<(), Error>> + Send>> {
```

File: client/service/src/lib.rs [hunk 7]
```diff
@@ -92,13 +93,13 @@ pub struct Service<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc> {
 	/// A receiver for spawned essential-tasks concluding.
 	essential_failed_rx: mpsc::UnboundedReceiver<()>,
 	/// Sender for futures that must be spawned as background tasks.
-	to_spawn_tx: mpsc::UnboundedSender<Box<dyn Future<Item = (), Error = ()> + Send>>,
+	to_spawn_tx: mpsc::UnboundedSender<Box<dyn Future<Output = ()> + Send + Unpin>>,
```
Comment (tomaka) [inline client/service/src/lib.rs]:
```suggestion
	to_spawn_tx: mpsc::UnboundedSender<Pin<Box<dyn Future<Output = ()> + Send>>>,
```

File: client/service/src/lib.rs [hunk 8]
```diff
@@ -92,13 +93,13 @@ pub struct Service<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc> {
...
 	/// Receiver for futures that must be spawned as background tasks.
-	to_spawn_rx: mpsc::UnboundedReceiver<Box<dyn Future<Item = (), Error = ()> + Send>>,
+	to_spawn_rx: mpsc::UnboundedReceiver<Box<dyn Future<Output = ()> + Send + Unpin>>,
```
Comment (tomaka) [inline client/service/src/lib.rs]: I'm not going to comment on every single one, but you get the idea. `Box<... + Unpin>` should be replaced with `Pin<Box<...>>`.

File: client/service/src/lib.rs [hunk 9]
```diff
@@ -109,42 +110,36 @@ pub struct Service<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc> {
...
+		self.spawn(future.compat().map(drop));
+		Ok(())
 	}
 }
 
 /// Abstraction over a Substrate service.
-pub trait AbstractService: 'static + Future<Item = (), Error = Error> +
-	Executor<Box<dyn Future<Item = (), Error = ()> + Send>> + Send {
+pub trait AbstractService: 'static + Future<Output = Result<(), Error>> +
+	Spawn + Send + Unpin {
```
Comment (tomaka) [inline client/service/src/lib.rs]:
```suggestion
	Spawn + Send {
```

File: client/service/src/lib.rs [hunk 10]
```diff
@@ -275,8 +267,12 @@ where
 		}
 	}
 
-	fn rpc_query(&self, mem: &RpcSession, request: &str) -> Box<dyn Future<Item = Option<String>, Error = ()> + Send> {
-		Box::new(self.rpc_handlers.handle_request(request, mem.metadata.clone()))
+	fn rpc_query(&self, mem: &RpcSession, request: &str) -> Box<dyn Future<Output = Option<String>> + Send + Unpin> {
```
Comment (tomaka) [inline client/service/src/lib.rs]:
```suggestion
	fn rpc_query(&self, mem: &RpcSession, request: &str) -> Pin<Box<dyn Future<Output = Option<String>> + Send>> {
```

File: client/service/src/status_sinks.rs [hunk 11]
```diff
@@ -57,50 +57,49 @@ impl<T> StatusSinks<T> {
...
+				Poll::Ready(Some((sender, interval))) => {
 					let status = status_grab();
 					if sender.unbounded_send(status).is_ok() {
 						self.entries.push(YieldAfter {
 							// Note that since there's a small delay between the moment a task is
 							// waken up and the moment it is polled, the period is actually not
 							// `interval` but `interval + <delay>`. We ignore this problem in
 							// practice.
-							delay: Delay::new(Instant::now() + interval),
+							delay: Delay::new( interval),
```
Comment (tomaka) [inline client/service/src/status_sinks.rs]:
```suggestion
							delay: Delay::new(interval),
```

File: client/consensus/aura/src/lib.rs [hunk 12]
```diff
@@ -994,18 +994,21 @@ mod tests {
...
+				.expect("Starts aura")
+				.unit_error()
+				.compat();
 
 			runtime.spawn(aura);
 		}
 
-		runtime.spawn(futures01::future::poll_fn(move || {
+		runtime.spawn(futures::future::poll_fn(move |_| {
 			net.lock().poll();
```
Comment (tomaka) [inline client/consensus/aura/src/lib.rs]: I don't think that's correct. The call to `poll()` here expects a futures 0.1 task available and will likely panic here?

File: client/consensus/babe/src/tests.rs [hunk 13]
```diff
@@ -403,16 +403,16 @@ fn run_one_test(
 			babe_link: data.link.clone(),
 			keystore,
 			can_author_with: sp_consensus::AlwaysCanAuthor,
-		}).expect("Starts babe"));
+		}).expect("Starts babe").unit_error().compat());
 	}
 
-	runtime.spawn(futures01::future::poll_fn(move || {
+	runtime.spawn(futures::future::poll_fn(move |_| {
 		net.lock().poll();
```
Comment (tomaka) [inline client/consensus/babe/src/tests.rs]: Same

File: client/consensus/aura/src/lib.rs (hunk 12 above)
Comment (expenses) [reply client/consensus/aura/src/lib.rs]: Well, the tests pass, so I'm not sure.

**Comment by expenses** [comment]
`cargo test -p node-cli --release --verbose --locked -- --ignored` does _eventually_ pass on my machine, so I'm not entirely sure what to do here.

**Comment by bkchr** [comment]
> `cargo test -p node-cli --release --verbose --locked -- --ignored` does _eventually_ pass on my machine, so I'm not entirely sure what to do here.

Modify the CI script to print the messages directly.

**Comment by gavofyork** [comment]
conflicts and CI failing.

File: client/service/src/lib.rs [hunk 14]
```diff
@@ -216,11 +210,11 @@ impl<TBl, TBackend, TExec, TRtApi, TSc, TNetSpec, TExPool, TOc> AbstractService
 	Service<TBl, Client<TBackend, TExec, TBl, TRtApi>, TSc, NetworkStatus<TBl>,
 		NetworkService<TBl, TNetSpec, TBl::Hash>, TExPool, TOc>
 where
-	TBl: BlockT,
+	TBl: BlockT + Unpin,
```
Comment (tomaka) [inline client/service/src/lib.rs]: This is wrong.

File: client/service/src/lib.rs [hunk 15]
```diff
@@ -309,57 +304,56 @@ where
 	}
 }
 
-impl<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc> Future for
+impl<TBl: Unpin, TCl, TSc: Unpin, TNetStatus, TNet, TTxPool, TOc> Future for
```
Comment (tomaka) [inline client/service/src/lib.rs]:
```suggestion
impl<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc> Future for
```

File: client/service/src/lib.rs [hunk 16]
```diff
@@ -216,11 +210,11 @@ impl<TBl, TBackend, TExec, TRtApi, TSc, TNetSpec, TExPool, TOc> AbstractService
 	Service<TBl, Client<TBackend, TExec, TBl, TRtApi>, TSc, NetworkStatus<TBl>,
 		NetworkService<TBl, TNetSpec, TBl::Hash>, TExPool, TOc>
 where
-	TBl: BlockT,
+	TBl: BlockT + Unpin,
 	TBackend: 'static + sc_client_api::backend::Backend<TBl>,
 	TExec: 'static + sc_client::CallExecutor<TBl> + Send + Sync + Clone,
 	TRtApi: 'static + Send + Sync,
-	TSc: sp_consensus::SelectChain<TBl> + 'static + Clone + Send,
+	TSc: sp_consensus::SelectChain<TBl> + 'static + Clone + Send + Unpin,
```
Comment (tomaka) [inline client/service/src/lib.rs]:
```suggestion
	TSc: sp_consensus::SelectChain<TBl> + 'static + Clone + Send,
```

File: client/service/src/lib.rs [hunk 17]
```diff
@@ -309,57 +304,56 @@ where
...
 				// Ready(None) should not be possible since we hold a live
 				// sender.
-				return Err(Error::Other("Essential task failed.".into()));
+				return Poll::Ready(Err(Error::Other("Essential task failed.".into())));
 			}
 		}
 
-		while let Ok(Async::Ready(Some(task_to_spawn))) = self.to_spawn_rx.poll() {
+		while let Poll::Ready(Some(task_to_spawn)) = Pin::new(&mut this.to_spawn_rx).poll_next(cx) {
+			// TODO: Update to tokio 0.2 when libp2p get switched to std futures (#4383)
```
Comment (tomaka) [inline client/service/src/lib.rs]: Libp2p has been switched to std futures.

**Comment by tomaka** [comment]
When you push 17 more commits after a previous review, I'd suggest to ask for a new review instead of just merging.

**Pull request merged**